lumeer-core/src/main/java/io/lumeer/core/task/executor/BlocklyRuleTaskExecutor.java

```diff
@@ -0,0 +1,319 @@
+
+   private static Logger log = Logger.getLogger(BlocklyRuleTaskExecutor.class.getName());
+
+   private String ruleName;
+   private BlocklyRule rule;
+   private RuleTask ruleTask;
+
+   public static class LumeerBridge {
+
+      private RuleTask ruleTask;
+      private Set<DocumentChange> changes = new HashSet<>();
+
+      private LumeerBridge(final RuleTask ruleTask) {
+         this.ruleTask = ruleTask;
+      }
+
+      public void setDocumentAttribute(DocumentBridge d, String attrId, Value value) {
+         changes.add(new DocumentChange(d.document, attrId, convertValue(value)));
+      }
+
+      Object convertValue(final Value value) {
+         if (value.isNumber()) {
+            return value.fitsInLong() ? value.asLong() : value.asDouble();
+         } else if (value.isBoolean()) {
+            return value.asBoolean();
+         } else if (value.isNull()) {
+            return null;
+         } else if (value.hasArrayElements()) {
+            final List list = new ArrayList();
+            for (long i = 0; i < value.getArraySize(); i = i + 1) {
+               list.add(convertValue(value.getArrayElement(i)));
+            }
+            return list;
+         } else {
+            return value.asString();
+         }
+      }
+
+      public List<DocumentBridge> getLinkedDocuments(DocumentBridge d, String linkTypeId) {
+         final SearchQuery query = SearchQuery
+               .createBuilder()
+               .stems(Arrays.asList(
+                     SearchQueryStem
+                           .createBuilder("")
+                           .linkTypeIds(Arrays.asList(linkTypeId))
+                           .documentIds(Set.of(d.document.getId()))
+                           .build()))
+               .build();
+
+         // load linked document ids
+         final Set<String> documentIds = new HashSet<>();
+         ruleTask.getDaoContextSnapshot().getLinkInstanceDao()
+                 .searchLinkInstances(query)
+                 .forEach(linkInstance -> {
+                    final List<String> documents = linkInstance.getDocumentIds();
+                    if (documents.size() >= 2) {
+                       if (documents.get(0).equals(d.document.getId())) {
+                          documentIds.add(documents.get(1));
+                       } else {
+                          documentIds.add(documents.get(0));
+                       }
+                    }
+                 });
+
+         // load document data
+         final Map<String, DataDocument> data = new HashMap<>();
+         ruleTask.getDaoContextSnapshot().getDataDao()
+                 .getData(ruleTask.getCollection().getId(), documentIds)
+                 .forEach(dd -> data.put(dd.getId(), dd));
+
+         // load document meta data and match them with user data
+         return ruleTask.getDaoContextSnapshot().getDocumentDao()
+                        .getDocumentsByIds(documentIds.toArray(new String[documentIds.size()]))
```

Comment:
It's better to use `new String[0]` as warning on this line suggests.

Reply:
I do not see any reason. According to the method implementation, if I pass an array to which the collection fits, it is just fine.